File: python/controller/controlador.py
from controller.conexion_bd import obtener_conexion
from model.guerrero import Guerrero
from model.mago import Mago
from model.arquero import Arquero

def instanciar_personaje(row):
   
    id_p, nombre, clase, nivel, _ = row
    if clase == "Guerrero":
        return Guerrero(nombre, nivel, id_p)
    elif clase == "Mago":
        return Mago(nombre, nivel, id_p)
    elif clase == "Arquero":
        return Arquero(nombre, nivel, id_p)
    return None

def crear_personaje(nombre, clase, nivel):
    conn = obtener_conexion()
    cursor = conn.cursor()
    query = "INSERT INTO personajes (nombre, clase, nivel, fecha_creacion) VALUES (%s, %s, %s, NOW())"
    cursor.execute(query, (nombre, clase, nivel))
    conn.commit()
    cursor.close()
    conn.close()

def listar_personajes():
    conn = obtener_conexion()
    cursor = conn.cursor()
    cursor.execute("SELECT id, nombre, clase, nivel, fecha_creacion FROM personajes")
    rows = cursor.fetchall()
    personajes = [instanciar_personaje(row) for row in rows]
    cursor.close()
    conn.close()
    return personajes

def buscar_por_nombre(nombre):
    conn = obtener_conexion()
    cursor = conn.cursor()
    query = "SELECT id, nombre, clase, nivel, fecha_creacion FROM personajes WHERE nombre = %s"
    cursor.execute(query, (nombre,))
    row = cursor.fetchone()
    cursor.close()
    conn.close()
    if row:
        return instanciar_personaje(row)
    return None

def actualizar_nivel(id_personaje, nuevo_nivel):
    conn = obtener_conexion()
    cursor = conn.cursor()
    
    # Validación de ID inexistente
    cursor.execute("SELECT id FROM personajes WHERE id = %s", (id_personaje,))
    if not cursor.fetchone():
        cursor.close()
        conn.close()
        return False
        
    query = "UPDATE personajes SET nivel = %s WHERE id = %s"
    cursor.execute(query, (nuevo_nivel, id_personaje))
    conn.commit()
    cursor.close()
    conn.close()
    return True

def eliminar_personaje(id_personaje):
    conn = obtener_conexion()
    cursor = conn.cursor()
    
    # Validación de ID inexistente
    cursor.execute("SELECT id FROM personajes WHERE id = %s", (id_personaje,))
    if not cursor.fetchone():
        cursor.close()
        conn.close()
        return False
        
    query = "DELETE FROM personajes WHERE id = %s"
    cursor.execute(query, (id_personaje,))
    conn.commit()
    cursor.close()
    conn.close()
    return True

File: java/src/controller/Controlador.java
package controller;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.List;
import java.util.ArrayList;
import model.Arquero;
import model.Guerrero;
import model.Mago;
import model.Personaje;

public class Controlador {
    public boolean crearPersonaje(String nombre, String clase, int nivel) {
        String sql = "INSERT INTO personajes (nombre, clase, nivel, fecha_creacion) VALUES (?, ?, ?, NOW())";

        try (Connection conn = ConexionBD.obtenerConexion();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setString(1, nombre);
            stmt.setString(2, normalizarClase(clase));
            stmt.setInt(3, nivel);
            stmt.executeUpdate();
            return true;
        } catch (SQLException e) {
            System.out.println("Error al crear personaje en MySQL: " + e.getMessage());
            return false;
        }
    }

    public List<Personaje> listarPersonajes() {
        List<Personaje> personajes = new ArrayList<Personaje>();
        String sql = "SELECT id, nombre, clase, nivel, fecha_creacion FROM personajes ORDER BY id";

        try (Connection conn = ConexionBD.obtenerConexion();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {
            while (rs.next()) {
                personajes.add(instanciarPersonaje(rs));
            }
        } catch (SQLException e) {
            System.out.println("Error al listar personajes desde MySQL: " + e.getMessage());
        }

        return personajes;
    }

    public Personaje buscarPorNombre(String nombre) {
        String sql = "SELECT id, nombre, clase, nivel, fecha_creacion FROM personajes WHERE nombre = ?";

        try (Connection conn = ConexionBD.obtenerConexion();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setString(1, nombre);

            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return instanciarPersonaje(rs);
                }
            }
        } catch (SQLException e) {
            System.out.println("Error al buscar personaje en MySQL: " + e.getMessage());
        }

        return null;
    }

    public boolean actualizarNivel(int idPersonaje, int nuevoNivel) {
        String sql = "UPDATE personajes SET nivel = ? WHERE id = ?";

        try (Connection conn = ConexionBD.obtenerConexion();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setInt(1, nuevoNivel);
            stmt.setInt(2, idPersonaje);
            return stmt.executeUpdate() > 0;
        } catch (SQLException e) {
            System.out.println("Error al actualizar personaje en MySQL: " + e.getMessage());
            return false;
        }
    }

    public boolean eliminarPersonaje(int idPersonaje) {
        String sql = "DELETE FROM personajes WHERE id = ?";

        try (Connection conn = ConexionBD.obtenerConexion();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setInt(1, idPersonaje);
            return stmt.executeUpdate() > 0;
        } catch (SQLException e) {
            System.out.println("Error al eliminar personaje en MySQL: " + e.getMessage());
            return false;
        }
    }

    private Personaje instanciarPersonaje(ResultSet rs) throws SQLException {
        return instanciarPersonaje(
                rs.getInt("id"),
                rs.getString("nombre"),
                rs.getString("clase"),
                rs.getInt("nivel")
        );
    }

    private Personaje instanciarPersonaje(int id, String nombre, String clase, int nivel) {
        String claseNormalizada = normalizarClase(clase);

        if ("Guerrero".equals(claseNormalizada)) {
            return new Guerrero(id, nombre, nivel);
        }

        if ("Mago".equals(claseNormalizada)) {
            return new Mago(id, nombre, nivel);
        }

        if ("Arquero".equals(claseNormalizada)) {
            return new Arquero(id, nombre, nivel);
        }

        return new Personaje(id, nombre, clase, nivel);
    }

    private String normalizarClase(String clase) {
        if ("guerrero".equalsIgnoreCase(clase)) {
            return "Guerrero";
        }

        if ("mago".equalsIgnoreCase(clase)) {
            return "Mago";
        }

        if ("arquero".equalsIgnoreCase(clase)) {
            return "Arquero";
        }

        return clase;
    }
}
